File: lib/models/transaction.dart
enum TransactionType {
  income,
  expense,
  transfer,
}

class Transaction {
  final String id;
  final double amount;
  final TransactionType type;
  final String categoryId;
  final String accountId;
  final DateTime date;
  final String note;

  Transaction({
    required this.id,
    required this.amount,
    required this.type,
    required this.categoryId,
    required this.accountId,
    required this.date,
    this.note = '',
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'amount': amount,
      'type': type.name,
      'categoryId': categoryId,
      'accountId': accountId,
      'date': date.toIso8601String(),
      'note': note,
    };
  }

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String,
      amount: (json['amount'] as num).toDouble(),
      type: TransactionType.values.firstWhere(
        (e) => e.name == json['type'],
        orElse: () => TransactionType.expense,
      ),
      categoryId: json['categoryId'] as String,
      accountId: json['accountId'] as String,
      date: DateTime.parse(json['date'] as String),
      note: json['note'] as String? ?? '',
    );
  }

  Transaction copyWith({
    String? id,
    double? amount,
    TransactionType? type,
    String? categoryId,
    String? accountId,
    DateTime? date,
    String? note,
  }) {
    return Transaction(
      id: id ?? this.id,
      amount: amount ?? this.amount,
      type: type ?? this.type,
      categoryId: categoryId ?? this.categoryId,
      accountId: accountId ?? this.accountId,
      date: date ?? this.date,
      note: note ?? this.note,
    );
  }
}

File: src/models/Transaction.ts
export enum TransactionType {
  INCOME = 'income',
  EXPENSE = 'expense',
  TRANSFER = 'transfer',
}

export interface Transaction {
  id: string;
  amount: number;
  type: TransactionType;
  categoryId: string;
  accountId: string;
  date: Date;
  note: string;
}

export const createTransaction = (
  id: string,
  amount: number,
  type: TransactionType,
  categoryId: string,
  accountId: string,
  date: Date,
  note: string = ''
): Transaction => ({
  id,
  amount,
  type,
  categoryId,
  accountId,
  date,
  note,
});

export const transactionToJson = (transaction: Transaction) => ({
  id: transaction.id,
  amount: transaction.amount,
  type: transaction.type,
  categoryId: transaction.categoryId,
  accountId: transaction.accountId,
  date: transaction.date.toISOString(),
  note: transaction.note,
});

export const transactionFromJson = (json: any): Transaction => ({
  id: json.id as string,
  amount: parseFloat(json.amount),
  type: json.type as TransactionType,
  categoryId: json.categoryId as string,
  accountId: json.accountId as string,
  date: new Date(json.date),
  note: json.note || '',
});

